solr/core/src/java/org/apache/solr/search/QueryUtils.java

```diff
@@ -54,23 +54,7 @@ public static boolean isNegative(Query q) {
    * varying score (i.e., it is a constant score query).
    */
   public static boolean isConstantScoreQuery(Query q) {
-    return isConstantScoreQuery(q, null);
-  }
-
-  private static Map<Query, Void> lazyInitSeen(Map<Query, Void> seen, Query add) {
-    if (seen == null) {
-      seen = new IdentityHashMap<>();
-    }
-    seen.put(add, null);
-    return seen;
-  }
-
-  /**
-   * Returns true if the specified query is guaranteed to assign the same score to all docs; otherwise false
-   * @param q query to be evaluated
-   * @param seen used to detect possible loops in nested query input
-   */
-  private static boolean isConstantScoreQuery(Query q, Map<Query, Void> seen) {
+    Map<Query, Void> seen = null; // lazy-init; this will be unnecessary in many cases
```

Comment:
Oops, was adding to `seen` but never checking it. The purpose should be clarified by 6074c5dc80a1984de74254b88fa7a748c1624735.

It is possible in theory to create a `WrappedQuery` that at some level wraps itself (perhaps indirectly). I'd be more than happy to get rid of this ... do you think it would be safe to do so?

Reply:
Yes definitely; it's simply wrong to create a circular reference in the first place.  It's very unreasonable to expect every consumer to have to check for this.

Reply:
This comment is also relevant:
```
// NOTE: Neither WrappedQuery class nor its inner query are final, so there is a risk of direct loops
// TODO: Only the queries we explicitly check for in this method are relevant wrt detecting loops, and
//  only `WrappedQuery` currently presents a risk in that respect; we may be able to avoid this risk
//  by more tightly restricting the `WrappedQuery` API (e.g., making the get/set methods `final`)?
```

Reply:
I'd remove these comments as well.  I've never seen this loop checking done before for Queries and I've never seen a Query try to prevent this.